### ISSUE TITLE
TODO: implement exact replication of bindcraft optimizer

### ISSUE THREAD

**Issue by nboyd** [escalante-bio/mosaic#35]

**Comment by nboyd** [comment]
hi @etherealsunshine , sure, sounds like a great PR! I wasn't thinking of bitwise parity with BindCraft (this would be very difficult -- our implementations of various loss terms probably differ slightly). Rather I was imagining a vanilla reimplementation of their 4stage optimizer; exactly what you're proposing! In theory this should work with any structure prediction model (af2,boltz, protenix...)

p.s. I haven't done this myself because I noticed various language model losses (ESMC PLL, proteinmpnn ll) do not work well with optimizers that go outside of the simplex. If you're up for it, it could be interesting to confirm that.

**Comment by etherealsunshine** [comment]
hi @nboyd, Thanks for the info! I did run a preliminary experiment based on the LM losses in logic space (and shared an email with more details).  

TLDR: LM-style sequence priors dont play well outside the simplex, which matches what you thought.

Just wanted to ask before i go ahead with implementation, what’d be best: 
1. **BindCraft-styled**: keep Stage 1 as unconstrained `design_logits`
2. **Mosaic-styled**: use simplex-native APGM for Stage 1

For Stages 2–4, they can be replicated w.r.t  ColabDesign-style behavior:
- Stage 2: soft mode
- Stage 3: hard mode
- Stage 4: PSSM semigreedy refinement

If we keep Stage 1 in logit space, I had a few ideas for the  LM loss issue :

- adaptive LM gradient budgeting,
- entropy floor / anti-saturation control,
- alternating LM/structure updates.

Do let me know what you think, thanks!
